### PR TITLE
[c10d] Make alltoall as a custom op

### PR DESCRIPTION
**Pull request by alanwaketan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79779
* #79777
* __->__ #79691

Summary:
This patch makes alltoall as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
BACKEND=nccl WORLD_SIZE=2 python test/distributed/test_distributed_spawn.py -v TestDistBackendWithSpawn.test_all_to_all_cuda
    BACKEND=nccl WORLD_SIZE=2 python test/distributed/test_distributed_spawn.py -v TestDistBackendWithSpawn.test_all_to_all_cuda_complex
    BACKEND=nccl WORLD_SIZE=2 python test/distributed/test_distributed_spawn.py -v TestDistBackendWithSpawn.test_all_to_all_full_group_cuda
    and other existing distributed tests.

### CODE REVIEW

File: torch/csrc/distributed/c10d/Ops.cpp [hunk 1]
```diff
@@ -97,6 +97,19 @@ c10::intrusive_ptr<ProcessGroup::Work> scatter_(
       ScatterOptions{root_rank, std::chrono::milliseconds(timeout)});
 }
 
+c10::intrusive_ptr<ProcessGroup::Work> alltoall_(
+    at::TensorList output_tensors,
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/Ops.cpp:101]: why the type here is different from other comm ops?

Comment (mrshenli) [reply torch/csrc/distributed/c10d/Ops.cpp:101]: not stamping yet due to this comment

Comment (alanwaketan) [reply torch/csrc/distributed/c10d/Ops.cpp:101]: I think it's a convention to use TensorList to represent const std::vector<at::Tensor>&.

I only use const std::vector<at::Tensor>& if there is a const std::vector<std::vector<at::Tensor>>& in the signature to keep consistency on that function. Let me know which way you like.

File: torch/csrc/distributed/c10d/init.cpp [hunk 2]
```diff
@@ -1263,18 +1263,25 @@ that adds a prefix to each key inserted to the store.
...
+              },
               py::arg("output_tensor"),
               py::arg("input_tensor"),
               py::arg("opts") = ::c10d::AllToAllOptions(),
               py::call_guard<py::gil_scoped_release>())
 
           .def(
               "alltoall",
-              [](::c10d::ProcessGroup& pg,
+              [](const c10::intrusive_ptr<::c10d::ProcessGroup>& pg,
```
[the remark this answers is not in the document]
Comment (alanwaketan) [reply torch/csrc/distributed/c10d/init.cpp:1280]: Let me follow up on https://github.com/pytorch/pytorch/issues/80246.

File: torch/csrc/distributed/c10d/Ops.hpp [hunk 3]
```diff
@@ -29,6 +29,10 @@ TORCH_API c10::intrusive_ptr<ProcessGroup::Work> scatter(const c10::intrusive_pt
       const std::vector<at::Tensor>& output_tensors,
       const std::vector<std::vector<at::Tensor>>& input_tensors,
       const ScatterOptions& opts ={});
+TORCH_API c10::intrusive_ptr<ProcessGroup::Work> alltoall(const c10::intrusive_ptr<ProcessGroup>& process_group,
+      at::TensorList output_tensors,
```
[the remark this answers is not in the document]
Comment (alanwaketan) [reply torch/csrc/distributed/c10d/Ops.hpp:33]: The pull request passes the linter. And the format seems consistent with other function signatures in the same file. What's your concern here?

File: torch/csrc/distributed/c10d/Ops.cpp (hunk 1 above)
Comment (mrshenli) [reply torch/csrc/distributed/c10d/Ops.cpp:101]: I see. can we add some docs to explain that in the code? Thanks!

File: torch/csrc/distributed/c10d/Ops.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (mrshenli) [reply torch/csrc/distributed/c10d/Ops.hpp:33]: We used to run clang-format on all distributed cpp files. Not sure what's today's convention. Below is what clang-format gives me:

```c++
namespace c10d {
namespace ops {

// Below are essentially ProcessGroup's corresponding ops but routed to the
// dispatcher.
TORCH_API c10::intrusive_ptr<ProcessGroup::Work>
broadcast(const c10::intrusive_ptr<ProcessGroup> &process_group,
          at::TensorList tensors, const BroadcastOptions &opts = {});
TORCH_API c10::intrusive_ptr<ProcessGroup::Work>
allreduce(const c10::intrusive_ptr<ProcessGroup> &process_group,
          at::TensorList tensors, const AllreduceOptions &opts = {});
TORCH_API c10::intrusive_ptr<ProcessGroup::Work>
allgather(const c10::intrusive_ptr<ProcessGroup> &process_group,
          const std::vector<std::vector<at::Tensor>> &output_tensors,
          const std::vector<at::Tensor> &input_tensors,
          const AllgatherOptions &opts = {});
TORCH_API c10::intrusive_ptr<ProcessGroup::Work>
reduce_scatter(const c10::intrusive_ptr<ProcessGroup> &process_group,
               const std::vector<at::Tensor> &output_tensors,
               const std::vector<std::vector<at::Tensor>> &input_tensors,
               const ReduceScatterOptions &opts = {});
TORCH_API c10::intrusive_ptr<ProcessGroup::Work>
reduce(const c10::intrusive_ptr<ProcessGroup> &process_group,
       at::TensorList tensors, const ReduceOptions &opts = {});
TORCH_API c10::intrusive_ptr<ProcessGroup::Work>
gather(const c10::intrusive_ptr<ProcessGroup> &process_group,
       const std::vector<std::vector<at::Tensor>> &output_tensors,
       const std::vector<at::Tensor> &input_tensors,
       const GatherOptions &opts = {});
TORCH_API c10::intrusive_ptr<ProcessGroup::Work>
scatter(const c10::intrusive_ptr<ProcessGroup> &process_group,
        const std::vector<at::Tensor> &output_tensors,
        const std::vector<std::vector<at::Tensor>> &input_tensors,
        const ScatterOptions &opts = {});
TORCH_API c10::intrusive_ptr<ProcessGroup::Work>
alltoall(const c10::intrusive_ptr<ProcessGroup> &process_group,
         at::TensorList output_tensors, at::TensorList input_tensors,
         const AllToAllOptions &opts = {});

} // namespace ops
} // namespace c10d
```

**Review by mrshenli — APPROVED** [review]
LGTM. Left a minor comment

**Review by wanchaol — APPROVED** [review]

File: torch/csrc/distributed/c10d/Ops.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (alanwaketan) [reply torch/csrc/distributed/c10d/Ops.hpp:33]: Somehow, my local linter doesn't produce the same output as yours. Let me just copy yours.

**Comment by alanwaketan** [comment]
Thanks for approving this pull request, Shen and Wanchao.

**Pull request closed without merging**